### PR TITLE
FairMQ PluginServices patches and various other

### PR DESCRIPTION
**Pull request by dennisklein**
see #561 for Todo list, focussing on high priority items in this PR.

### CODE REVIEW

**Comment by dennisklein** [comment]
@rbx I rebased this PR to current `dev` and added the static plugin mechanism commit. This should be a good starting point, the diff of d4615a1 should be self explaining. I tried with header only plugin, but the compiler complained about multiple definitions of the prog options function, so I moved it to the cxx file.

```bash
cmake --build . --target FairMQTests -- -j5
bin/testsuite_FairMQ.Plugins --gtest_filter="*LoadPluginStatic*"
```
The above is the easiest way to run the `control_static` plugin at the moment. Of course, we should add at least another test case, because this one is mainly about the `LoadPluginStatic` method of the `PluginManager`.

**Comment by dennisklein** [comment]
@rbx I added a first version of the builtin control plugin in 41b5adc.

```bash
cmake --build . --target bsampler -- -j5
bin/bsampler --id bsampler1 --mq-config bin/config/benchmark.json --control static
```
Run the above and then press Ctrl-C, there is still a deadlock on `ReleaseDeviceControl()`. Will investigate tomorrow.

I have not touched interactive mode yet. Probably need to move the code from device to the plugin, because I can't call `InteractiveStateLoop()` directly.

You can see my draft how `runFairMQDevice.h` no longer includes `runSimpleMQStateMachine.h` and how it looks like now.

Also, what do you think about moving the `--control` option away from central ProgOptions to the control plugin specific options (where I put now `--ctrlmode` as a placeholder)?

**Comment by rbx** [comment]
Regarding the runSimpleMQStateMachine.h, there is at least one more (recent) user of it here:
https://github.com/AliceO2Group/AliceO2/blob/dev/Framework/Core/src/runDataProcessing.cxx#L247

I do like how runFairMQDevice looks like in your branch. But thinking about it, maybe it is a good idea to keep this abstraction (at least the runStateMachine function, and perhaps not the header file). We provide a common main() function for many users. But for advanced use cases, such as above, hiding our details in something like `runStateMachine` or `runDevice` still has benefit for the advanced users.

Moving the `--control` option is fine with me.

File: fairmq/PluginServices.cxx [hunk 1]
```diff
@@ -85,3 +86,81 @@ const std::unordered_map<PluginServices::DeviceStateTransition, FairMQDevice::Ev
...
+    lock_guard<mutex> lock{fDeviceControllerMutex};
+
+    return fDeviceController;
+}
+
+auto PluginServices::WaitForReleaseDeviceControl() -> void
+{
+    unique_lock<mutex> lock{fDeviceControllerMutex};
+
+    while(GetDeviceController())
```
Comment (rbx) [inline fairmq/PluginServices.cxx]: The deadlock happens here - you are locking `fDeviceControllerMutex`, then calling `GetDeviceController`, which tries to lock the same mutex again. I fixed it by just replacing call to `GetDeviceController` with `fDeviceController`. Don't really need the GetDeviceController method, unless you want to use it from somewhere else too.

**Comment by dennisklein** [comment]
@rbx Merged, thx :)

**Comment by dennisklein** [comment]
@rbx Noticed something odd, but haven't investigated yet:

```bash
╭─dklein@gamma ~/FairRoot2/build ‹ruby-2.3.4› ‹plugins2› 
╰─➤  bin/bsampler --id bsampler1 --channel-config name=data,method=bind,type=pub,address=tcp://127.0.0.1:5555 --control static --verbosity INFO
[02:20:57][STATE] Starting FairMQ state machine
[02:20:57][STATE] Entering INITIALIZING DEVICE state
[02:20:57][STATE] Entering DEVICE READY state
[02:20:57][STATE] Entering INITIALIZING TASK state
[02:20:57][STATE] Entering READY state
[02:20:57][STATE] Entering RUNNING state
[02:20:57][INFO] DEVICE: Running...
[02:20:57][INFO] Starting the benchmark with message size of 1000 and 0 iterations.
^C[02:20:59][INFO] Received device shutdown request (signal 2).
[02:20:59][INFO] Waiting for graceful device shutdown. Hit Ctrl-C again to abort immediately.
[02:20:59][STATE] Entering READY state
[02:20:59][INFO] Leaving RUNNING state. Done 12100351 iterations in 1880.79ms.
[02:20:59][STATE] Entering RESETTING TASK state
[02:20:59][STATE] Entering DEVICE READY state
[02:20:59][STATE] Entering RESETTING DEVICE state
[02:20:59][STATE] Entering IDLE state
[02:20:59][STATE] Entering EXITING state
[02:20:59][STATE] Exiting FairMQ state machine
```

```bash
╭─dklein@gamma ~/FairRoot2/build ‹ruby-2.3.4› ‹plugins2› 
╰─➤  bin/bsampler --id bsampler1 --channel-config name=data,method=bind,type=pub,address=tcp://127.0.0.1:5555 --control interactive --verbosity INFO
[02:21:08][STATE] Starting FairMQ state machine
[02:21:08][STATE] Entering INITIALIZING DEVICE state
[02:21:08][STATE] Entering DEVICE READY state
[02:21:08][STATE] Entering INITIALIZING TASK state
[02:21:08][STATE] Entering READY state
[02:21:08][STATE] Entering RUNNING state
[02:21:08][INFO] DEVICE: Running...
[02:21:08][INFO] Use keys to control the state machine:
[02:21:08][INFO] [h] help, [p] pause, [r] run, [s] stop, [t] reset task, [d] reset device, [q] end, [j] init task, [i] init device
[02:21:08][INFO] Starting the benchmark with message size of 1000 and 0 iterations.
q[02:21:09][INFO] [q] end
[02:21:09][STATE] Entering READY state
[02:21:09][INFO] Leaving RUNNING state. Done 11237399 iterations in 1738.42ms.
[02:21:10][STATE] Entering RESETTING TASK state
[02:21:10][STATE] Entering DEVICE READY state
[02:21:10][STATE] Entering RESETTING DEVICE state
[02:21:10][STATE] Entering IDLE state
[02:21:10][STATE] Entering EXITING state
[02:21:10][STATE] Exiting FairMQ state machine
```

`Entering READY state` is printed too early, isn't it?

**Comment by rbx** [comment]
The `[INFO]` is printed by the bsampler code (not the state machine), after the state has changed from RUNNING and it is going out of Run(). i'll modify it to print something less misleading.

**Comment by dennisklein** [comment]
> The [INFO] is printed by the bsampler code (not the state machine), after the state has changed from RUNNING and it is going out of Run(). i'll modify it to print something not misleading.

See 3ac3ef4, I believe that reflects more correctly to the user, when the new state starts.

**Comment by dennisklein** [comment]
Hm, I guess we should wait for the current job in the worker thread in the state the job was queued. The way it is done now, I can call ChangeState in my device code and then get wrong results when asking for current state after the ChangeState call (it will already report the next state, won't it?).

**Comment by rbx** [comment]
But that would prevent something like `while(CheckCurrentState(RUNNING)) {...}` from working. Or do I misunderstand?

**Comment by dennisklein** [comment]
So, currently, I make a subclass of FairMQDevice and implement Run method. That code in the Run method will not run in RUNNING state only. That was my expectation up to now.

**Comment by rbx** [comment]
The Run method needs some kind of stop condition, which is currently end of the RUNNING state.

**Comment by dennisklein** [comment]
Sure, I still expected that the actual state transition will only happen after Run finished.

**Comment by dennisklein** [comment]
Let's say plugins subscribe to state changes, they will get notified too early.

**Comment by rbx** [comment]
The callbacks are called from the worker thread, after work (e.g. Run()) is completed.

**Comment by dennisklein** [comment]
Hm, can we move the `fsm.fState` assignment also after the worker waiting code like I did with the LOG lines?

**Comment by rbx** [comment]
Only if we change the stop condition of the Run() handler from `CheckCurrentState(RUNNING)` to something else. Otherwise it will deadlock - it will return true until state is changed, and the state wont be changed until Run stops.

**Comment by dennisklein** [comment]
If we cant/dont want to do that, I see only the option to remove GetCurrentDeviceState from plugin api. Because currently it is broken. Or I add a second state member which reflects the correct state.

**Comment by rbx** [comment]
The subscriptions are called only after the state handlers are finished, therefore plugins cannot get notified too early. Where is it broken?

**Comment by dennisklein** [comment]
`GetCurrentDeviceState()` will return next state after change state call, won't it? Which is not true, if the current worker has not finished yet.

**Comment by rbx** [comment]
Now I see the issue, agreed. Changing the stop condition would be the cleanest, since it anyway only has a meaningful use in the Run(). But it is used a lot. Let's deprecate it and make a new stop condition available. For the plugin we make a second state member, until the original is fixed. What do you think?

**Comment by dennisklein** [comment]
@rbx If you are happy with 351839f I would rip the old one out and see how it goes.

**Comment by rbx** [comment]
Need to actually hook it up to the device and test to be sure. I'm not convinced we should really get rid of MSM. I think the tree other points (simplify the code, drop threaded execution, and transitions queuing) can be achieved while keeping the MSM.

**Comment by dennisklein** [comment]
@rbx This test is hanging. I will debug, but maybe you know it instantly, when you see it.

```
[ RUN      ] Poller.ShMem_subchannel
[POLLOUT]/home/dklein/FairRoot2/build/bin/testhelper_runTestDevice --id pollout_shmem --control static --verbosity DEBUG --log-color false --mq-config "/home/dklein/FairRoot2/build/bin/testsuite_FairMQ.IOPatterns_config.json"
[POLLIN]/home/dklein/FairRoot2/build/bin/testhelper_runTestDevice --id pollin_shmem --control static --verbosity DEBUG --log-color false --mq-config "/home/dklein/FairRoot2/build/bin/testsuite_FairMQ.IOPatterns_config.json" --poll-type 0
[POLLOUT][18:05:50][DEBUG] mq-config: Using default XML/JSON parser
[POLLOUT][18:05:50][DEBUG] Looking for 'pollout_shmem' id/key in the provided config file...
[POLLOUT][18:05:50][DEBUG] Found with following channels:
[POLLOUT][18:05:50][DEBUG] data1:
[POLLOUT][18:05:50][DEBUG] 	No sockets specified,
[POLLOUT][18:05:50][DEBUG] 	applying common settings to the channel:
[POLLOUT][18:05:50][DEBUG] 	type          = push
[POLLOUT][18:05:50][DEBUG] 	method        = bind
[POLLOUT][18:05:50][DEBUG] 	address       = tcp://127.0.0.1:6004
[POLLOUT][18:05:50][DEBUG] 	transport     = shmem
[POLLOUT][18:05:50][DEBUG] 	sndBufSize    = 1000
[POLLOUT][18:05:50][DEBUG] 	rcvBufSize    = 1000
[POLLOUT][18:05:50][DEBUG] 	sndKernelSize = 0
[POLLOUT][18:05:50][DEBUG] 	rcvKernelSize = 0
[POLLOUT][18:05:50][DEBUG] 	rateLogging   = 0
[POLLOUT][18:05:50][DEBUG] data2:
[POLLOUT][18:05:50][DEBUG] 	No sockets specified,
[POLLOUT][18:05:50][DEBUG] 	applying common settings to the channel:
[POLLOUT][18:05:50][DEBUG] 	type          = push
[POLLOUT][18:05:50][DEBUG] 	method        = bind
[POLLOUT][18:05:50][DEBUG] 	address       = tcp://127.0.0.1:6005
[POLLOUT][18:05:50][DEBUG] 	transport     = shmem
[POLLOUT][18:05:50][DEBUG] 	sndBufSize    = 1000
[POLLOUT][18:05:50][DEBUG] 	rcvBufSize    = 1000
[POLLOUT][18:05:50][DEBUG] 	sndKernelSize = 0
[POLLOUT][18:05:50][DEBUG] 	rcvKernelSize = 0
[POLLOUT][18:05:50][DEBUG] 	rateLogging   = 0
[POLLOUT][18:05:50][DEBUG] ******************************************************************************************************************************
[POLLOUT][18:05:50][DEBUG] ****************************************************     Configuration     ***************************************************
[POLLOUT][18:05:50][DEBUG] ******************************************************************************************************************************
[POLLOUT][18:05:50][DEBUG] catch-signals               = 1                                                                       <int>         [default] 
[POLLOUT][18:05:50][DEBUG] chans.data1.0.address       = tcp://127.0.0.1:6004                                                    <string>      [provided]
[POLLOUT][18:05:50][DEBUG] chans.data1.0.method        = bind                                                                    <string>      [provided]
[POLLOUT][18:05:50][DEBUG] chans.data1.0.rateLogging   = 0                                                                       <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data1.0.rcvBufSize    = 1000                                                                    <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data1.0.rcvKernelSize = 0                                                                       <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data1.0.sndBufSize    = 1000                                                                    <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data1.0.sndKernelSize = 0                                                                       <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data1.0.transport     = shmem                                                                   <string>      [provided]
[POLLOUT][18:05:50][DEBUG] chans.data1.0.type          = push                                                                    <string>      [provided]
[POLLOUT][18:05:50][DEBUG] chans.data2.0.address       = tcp://127.0.0.1:6005                                                    <string>      [provided]
[POLLOUT][18:05:50][DEBUG] chans.data2.0.method        = bind                                                                    <string>      [provided]
[POLLIN][18:05:50][DEBUG] mq-config: Using default XML/JSON parser
[POLLIN][18:05:50][DEBUG] Looking for 'pollin_shmem' id/key in the provided config file...
[POLLIN][18:05:50][DEBUG] Found with following channels:
[POLLIN][18:05:50][DEBUG] data1:
[POLLIN][18:05:50][DEBUG] 	No sockets specified,
[POLLIN][18:05:50][DEBUG] 	applying common settings to the channel:
[POLLIN][18:05:50][DEBUG] 	type          = pull
[POLLIN][18:05:50][DEBUG] 	method        = connect
[POLLIN][18:05:50][DEBUG] 	address       = tcp://127.0.0.1:6004
[POLLIN][18:05:50][DEBUG] 	transport     = shmem
[POLLIN][18:05:50][DEBUG] 	sndBufSize    = 1000
[POLLIN][18:05:50][DEBUG] 	rcvBufSize    = 1000
[POLLIN][18:05:50][DEBUG] 	sndKernelSize = 0
[POLLIN][18:05:50][DEBUG] 	rcvKernelSize = 0
[POLLIN][18:05:50][DEBUG] 	rateLogging   = 0
[POLLIN][18:05:50][DEBUG] data2:
[POLLIN][18:05:50][DEBUG] 	No sockets specified,
[POLLIN][18:05:50][DEBUG] 	applying common settings to the channel:
[POLLIN][18:05:50][DEBUG] 	type          = pull
[POLLIN][18:05:50][DEBUG] 	method        = connect
[POLLIN][18:05:50][DEBUG] 	address       = tcp://127.0.0.1:6005
[POLLIN][18:05:50][DEBUG] 	transport     = shmem
[POLLIN][18:05:50][DEBUG] 	sndBufSize    = 1000
[POLLIN][18:05:50][DEBUG] 	rcvBufSize    = 1000
[POLLIN][18:05:50][DEBUG] 	sndKernelSize = 0
[POLLIN][18:05:50][DEBUG] 	rcvKernelSize = 0
[POLLIN][18:05:50][DEBUG] 	rateLogging   = 0
[POLLIN][18:05:50][DEBUG] ******************************************************************************************************************************
[POLLIN][18:05:50][DEBUG] ****************************************************     Configuration     ***************************************************
[POLLIN][18:05:50][DEBUG] ******************************************************************************************************************************
[POLLIN][18:05:50][DEBUG] catch-signals               = 1                                                                       <int>         [default] 
[POLLIN][18:05:50][DEBUG] chans.data1.0.address       = tcp://127.0.0.1:6004                                                    <string>      [provided]
[POLLIN][18:05:50][DEBUG] chans.data1.0.method        = connect                                                                 <string>      [provided]
[POLLIN][18:05:50][DEBUG] chans.data1.0.rateLogging   = 0                                                                       <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data1.0.rcvBufSize    = 1000                                                                    <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data1.0.rcvKernelSize = 0                                                                       <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data1.0.sndBufSize    = 1000                                                                    <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data1.0.sndKernelSize = 0                                                                       <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data1.0.transport     = shmem                                                                   <string>      [provided]
[POLLIN][18:05:50][DEBUG] chans.data1.0.type          = pull                                                                    <string>      [provided]
[POLLIN][18:05:50][DEBUG] chans.data2.0.address       = tcp://127.0.0.1:6005                                                    <string>      [provided]
[POLLIN][18:05:50][DEBUG] chans.data2.0.method        = connect                                                                 <string>      [provided]
[POLLIN][18:05:50][DEBUG] chans.data2.0.rateLogging   = 0                                                                       <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data2.0.rcvBufSize    = 1000                                                                    <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data2.0.rcvKernelSize = 0                                                                       <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data2.0.sndBufSize    = 1000                                                                    <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data2.0.sndKernelSize = 0                                                                       <int>         [provided]
[POLLIN][18:05:50][DEBUG] chans.data2.0.transport     = shmem                                                                   <string>      [provided]
[POLLIN][18:05:50][DEBUG] chans.data2.0.type          = pull                                                                    <string>      [provided]
[POLLIN][18:05:50][DEBUG] config                      = static                                                                  <string>      [default] 
[POLLIN][18:05:50][DEBUG] control                     = static                                                                  <string>      [provided]
[POLLIN][18:05:50][DEBUG] data1.numSockets            = 1                                                                       <int>         [provided]
[POLLIN][18:05:50][DEBUG] data2.numSockets            = 1                                                                       <int>         [provided]
[POLLIN][18:05:50][DEBUG] id                          = pollin_shmem                                                            <string>      [provided]
[POLLIN][18:05:50][DEBUG] initialization-timeout      = 120                                                                     <int>         [default] 
[POLLIN][18:05:50][DEBUG] io-threads                  = 1                                                                       <int>         [default] 
[POLLIN][18:05:50][DEBUG] log-color                   = 0                                                                       <bool>        [provided]
[POLLIN][18:05:50][DEBUG] log-to-file                 =                                                                         <string>      [default] 
[POLLIN][18:05:50][DEBUG] mq-config                   = /home/dklein/FairRoot2/build/bin/testsuite_FairMQ.IOPatterns_config.json<string>      [provided]
[POLLIN][18:05:50][DEBUG] network-interface           = default                                                                 <string>      [default] 
[POLLIN][18:05:50][DEBUG] poll-type                   = 0                                                                       <int>         [provided]
[POLLIN][18:05:50][DEBUG] port-range-max              = 32000                                                                   <int>         [default] 
[POLLIN][18:05:50][DEBUG] port-range-min              = 22000                                                                   <int>         [default] 
[POLLIN][18:05:50][DEBUG] shm-segment-name            = fairmq_shmem_main                                                       <string>      [default] 
[POLLIN][18:05:50][DEBUG] shm-segment-size            = 2000000000                                                              <std::size_t> [default] 
[POLLIN][18:05:50][DEBUG] transport                   = zeromq                                                                  <string>      [default] 
[POLLIN][18:05:50][DEBUG] verbosity                   = DEBUG                                                                   <string>      [provided]
[POLLIN][18:05:50][DEBUG] ******************************************************************************************************************************
[POLLIN][18:05:50][STATE] Starting FairMQ state machine
[POLLIN][18:05:50][DEBUG] PID: 10453
[POLLIN][18:05:50][DEBUG] Requesting 'zeromq' as default transport for the device
[POLLIN][18:05:50][DEBUG] Transport: Using ZeroMQ library, version: 4.2.0
[POLLIN][18:05:50][DEBUG] Adding 'zeromq' transport to the device.
[POLLOUT][18:05:50][DEBUG] chans.data2.0.rateLogging   = 0                                                                       <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data2.0.rcvBufSize    = 1000                                                                    <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data2.0.rcvKernelSize = 0                                                                       <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data2.0.sndBufSize    = 1000                                                                    <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data2.0.sndKernelSize = 0                                                                       <int>         [provided]
[POLLOUT][18:05:50][DEBUG] chans.data2.0.transport     = shmem                                                                   <string>      [provided]
[POLLOUT][18:05:50][DEBUG] chans.data2.0.type          = push                                                                    <string>      [provided]
[POLLOUT][18:05:50][DEBUG] config                      = static                                                                  <string>      [default] 
[POLLOUT][18:05:50][DEBUG] control                     = static                                                                  <string>      [provided]
[POLLOUT][18:05:50][DEBUG] data1.numSockets            = 1                                                                       <int>         [provided]
[POLLOUT][18:05:50][DEBUG] data2.numSockets            = 1                                                                       <int>         [provided]
[POLLOUT][18:05:50][DEBUG] id                          = pollout_shmem                                                           <string>      [provided]
[POLLOUT][18:05:50][DEBUG] initialization-timeout      = 120                                                                     <int>         [default] 
[POLLOUT][18:05:50][DEBUG] io-threads                  = 1                                                                       <int>         [default] 
[POLLOUT][18:05:50][DEBUG] log-color                   = 0                                                                       <bool>        [provided]
[POLLOUT][18:05:50][DEBUG] log-to-file                 =                                                                         <string>      [default] 
[POLLOUT][18:05:50][DEBUG] mq-config                   = /home/dklein/FairRoot2/build/bin/testsuite_FairMQ.IOPatterns_config.json<string>      [provided]
[POLLOUT][18:05:50][DEBUG] network-interface           = default                                                                 <string>      [default] 
[POLLOUT][18:05:50][DEBUG] poll-type                   = 0                                                                       <int>         [default] 
[POLLOUT][18:05:50][DEBUG] port-range-max              = 32000                                                                   <int>         [default] 
[POLLOUT][18:05:50][DEBUG] port-range-min              = 22000                                                                   <int>         [default] 
[POLLOUT][18:05:50][DEBUG] shm-segment-name            = fairmq_shmem_main                                                       <string>      [default] 
[POLLOUT][18:05:50][DEBUG] shm-segment-size            = 2000000000                                                              <std::size_t> [default] 
[POLLOUT][18:05:50][DEBUG] transport                   = zeromq                                                                  <string>      [default] 
[POLLIN][18:05:50][DEBUG] Loaded plugin: 'control', version '1.0.1', maintainer 'FairRootGroup <fairroot@gsi.de>', homepage 'https://github.com/FairRootGroup/FairRoot'
[POLLOUT][18:05:50][DEBUG] verbosity                   = DEBUG                                                                   <string>      [provided]
[POLLIN][18:05:50][DEBUG] Running builtin controller: static
[POLLOUT][18:05:50][DEBUG] ******************************************************************************************************************************
[POLLOUT][18:05:50][STATE] Starting FairMQ state machine
[POLLIN][18:05:50][DEBUG] catch-signals: 1
[POLLIN][18:05:50][STATE] Entering INITIALIZING DEVICE state
[POLLOUT][18:05:50][DEBUG] PID: 10454
[POLLOUT][18:05:50][DEBUG] Requesting 'zeromq' as default transport for the device
[POLLOUT][18:05:50][DEBUG] Transport: Using ZeroMQ library, version: 4.2.0
[POLLOUT][18:05:50][DEBUG] Adding 'zeromq' transport to the device.
[POLLIN][18:05:50][DEBUG] Validating channel "data1[0]"... VALID
[POLLIN][18:05:50][DEBUG] data1[0]: channel transport (zeromq) overriden to shmem
[POLLOUT][18:05:50][DEBUG] Loaded plugin: 'control', version '1.0.1', maintainer 'FairRootGroup <fairroot@gsi.de>', homepage 'https://github.com/FairRootGroup/FairRoot'
[POLLOUT][18:05:50][DEBUG] Running builtin controller: static
[POLLOUT][18:05:50][DEBUG] catch-signals: 1
[POLLOUT][18:05:50][STATE] Entering INITIALIZING DEVICE state
[POLLOUT][18:05:50][DEBUG] Validating channel "data1[0]"... VALID
[POLLOUT][18:05:50][DEBUG] data1[0]: channel transport (zeromq) overriden to shmem
[POLLOUT][18:05:50][DEBUG] Transport: Using ZeroMQ (4.2.0) & boost::interprocess (1.64.0)
[POLLOUT][18:05:50][DEBUG] shmem: created/opened shared memory segment of 2000000000 bytes. Available are 1999999680 bytes.
[POLLOUT][18:05:50][DEBUG] shmem: device counter found, with value of 9. incrementing.
[POLLOUT][18:05:50][DEBUG] shmem: incremented device counter, now: 10
[POLLOUT][18:05:50][DEBUG] Adding 'shmem' transport to the device.
[POLLOUT][18:05:50][DEBUG] Could not bind to configured (TCP) port, trying random port in range 22000-32000
[POLLOUT][18:05:50][DEBUG] Attached channel data1[0] to tcp://127.0.0.1:28359 (bind) 
[POLLOUT][18:05:50][DEBUG] Validating channel "data2[0]"... VALID
[POLLOUT][18:05:50][DEBUG] data2[0]: channel transport (zeromq) overriden to shmem
[POLLOUT][18:05:50][DEBUG] Reusing existing 'shmem' transport.
[POLLOUT][18:05:50][DEBUG] Could not bind to configured (TCP) port, trying random port in range 22000-32000
[POLLOUT][18:05:50][DEBUG] Attached channel data2[0] to tcp://127.0.0.1:29728 (bind) 
[POLLIN][18:05:50][DEBUG] Transport: Using ZeroMQ (4.2.0) & boost::interprocess (1.64.0)
[POLLIN][18:05:50][DEBUG] shmem: created/opened shared memory segment of 2000000000 bytes. Available are 1999999680 bytes.
[POLLIN][18:05:50][DEBUG] shmem: device counter found, with value of 10. incrementing.
[POLLIN][18:05:50][DEBUG] shmem: incremented device counter, now: 11
[POLLIN][18:05:50][DEBUG] Adding 'shmem' transport to the device.
[POLLIN][18:05:50][DEBUG] Attached channel data1[0] to tcp://127.0.0.1:6004 (connect) 
[POLLIN][18:05:50][DEBUG] Validating channel "data2[0]"... VALID
[POLLIN][18:05:50][DEBUG] data2[0]: channel transport (zeromq) overriden to shmem
[POLLIN][18:05:50][DEBUG] Reusing existing 'shmem' transport.
[POLLIN][18:05:50][DEBUG] Attached channel data2[0] to tcp://127.0.0.1:6005 (connect) 
[POLLOUT][18:05:51][STATE] Entering DEVICE READY state
[POLLOUT][18:05:51][STATE] Entering INITIALIZING TASK state
[POLLOUT][18:05:51][STATE] Entering READY state
[POLLOUT][18:05:51][STATE] Entering RUNNING state
[POLLOUT][18:05:51][INFO] DEVICE: Running...
[POLLIN][18:05:51][STATE] Entering DEVICE READY state
[POLLIN][18:05:51][STATE] Entering INITIALIZING TASK state
[POLLIN][18:05:51][STATE] Entering READY state
[POLLIN][18:05:51][STATE] Entering RUNNING state
[POLLIN][18:05:51][INFO] DEVICE: Running...
```

**Comment by dennisklein** [comment]
@MohammadAlTurany This PR is now feature complete. Once we have fixed all failing tests, this can be merged (I will remove the "in progress" label when this is done).

What is the status of travis tests? Should they pass too or are they broken at the moment?

**Comment by dennisklein** [comment]
> Once we have fixed all failing tests, this can be merged (I will remove the "in progress" label when this is done).
Ah, and @rbx needs to approve the review too before we can merge.

**Comment by rbx** [comment]
@dennisklein the tests pass for me. it looks like you have some old processes running, one of devices fails to bind the port and picks a random one. And those running ones still hold shared memory, run also `shmonitor --cleanup` once. Maybe we should run it before/after shm tests. I do it in the multi-transport example, but not in bin/testsuite_FairMQ.Protocols .

**Comment by dennisklein** [comment]
@rbx thx! had in fact old processes running. So, the gtest based testsuites now all succeed for me. Will do a full compilation now and see if the examples work.

**Comment by rbx** [comment]
I have another test failing on linux (although it runs on MacOS):
```
[ RUN      ] PluginManager.LoadPluginDynamic
[19:09:20][STATE] Starting FairMQ state machine
/home/orybalch/dev/fair/base/fairmq/test/plugins/_plugin_manager.cxx:49: Failure
Expected: mgr.LoadPlugin("test_dummy") doesn't throw an exception.
  Actual: it throws.
[19:09:20][DEBUG] Destructing device 
[19:09:20][STATE] Exiting FairMQ state machine
terminate called without an active exception
Aborted (core dumped)
```

**Comment by dennisklein** [comment]
`gcc (Debian 4.9.2-10) 4.9.2` and `gcc (GCC) 7.1.1 20170622 (Red Hat 7.1.1-3)` successfully compile and run our full ctest suite without errors. I will test gcc 5 and 6 too, but needs a while, since i have not installed them anywhere currently.

**Comment by dennisklein** [comment]
Travis GCC4.8 is also happy, the MacOS travis test will fail because of FairSoft 16 I assume.

**Comment by rbx** [comment]
The exception that is thrown in the failing test (on `mgr.LoadPlugin("test_dummy")`) is:
```
locale::facet::_S_create_c_locale name not valid
```

**Comment by dennisklein** [comment]
@rbx Can you post the output of `env | grep "^LC"` run the shell you are running the tests in. Also, pls post the output of `apt-cache policy locales` and `locale -a`.Thx!

**Comment by dennisklein** [comment]
`LoadPlugin` uses `boost::filesystem::path` internally and this seems to have a dependency to the system locale regarding its encoding.

**Comment by rbx** [comment]
```bash
orybalch@depc437:~/dev/fair/base/build$ env | grep "^LC"
LC_MEASUREMENT=de_DE.UTF-8
LC_PAPER=de_DE.UTF-8
LC_MONETARY=de_DE.UTF-8
LC_NAME=de_DE.UTF-8
LC_CTYPE=UTF-8
LC_ADDRESS=de_DE.UTF-8
LC_NUMERIC=de_DE.UTF-8
LC_TELEPHONE=de_DE.UTF-8
LC_IDENTIFICATION=de_DE.UTF-8
LC_TIME=de_DE.UTF-8
orybalch@depc437:~/dev/fair/base/build$ apt-cache policy locales
locales:
  Installed: 2.24-9ubuntu2.2
  Candidate: 2.24-9ubuntu2.2
  Version table:
 *** 2.24-9ubuntu2.2 500
        500 http://de.archive.ubuntu.com/ubuntu zesty-updates/main amd64 Packages
        500 http://de.archive.ubuntu.com/ubuntu zesty-updates/main i386 Packages
        500 http://security.ubuntu.com/ubuntu zesty-security/main amd64 Packages
        500 http://security.ubuntu.com/ubuntu zesty-security/main i386 Packages
        100 /var/lib/dpkg/status
     2.24-9ubuntu2 500
        500 http://de.archive.ubuntu.com/ubuntu zesty/main amd64 Packages
        500 http://de.archive.ubuntu.com/ubuntu zesty/main i386 Packages
orybalch@depc437:~/dev/fair/base/build$ locale -a
locale: Cannot set LC_CTYPE to default locale: No such file or directory
C
C.UTF-8
de_DE.utf8
en_AG
en_AG.utf8
en_AU.utf8
en_BW.utf8
en_CA.utf8
en_DK.utf8
en_GB.utf8
en_HK.utf8
en_IE.utf8
en_IL
en_IL.utf8
en_IN
en_IN.utf8
en_NG
en_NG.utf8
en_NZ.utf8
en_PH.utf8
en_SG.utf8
en_US.utf8
en_ZA.utf8
en_ZM
en_ZM.utf8
en_ZW.utf8
POSIX
orybalch@depc437:~/dev/fair/base/build$ 
```

**Comment by rbx** [comment]
After adding `export LC_ALL="en_US.UTF-8"` to my `.bashrc` file, tests run now.

**Comment by dennisklein** [comment]
Cool. So `LC_ALL` seems to override `LC_CTYPE`? I would guess `LC_CTYPE` must be `C` or `C.UTF-8` to also get rid of the error `locale -a` throws.

**Comment by rbx** [comment]
Yeah, after setting LC_ALL, LC_CTYPE also has the same value.

**Comment by dennisklein** [comment]
@MohammadAlTurany fixed the regression bug and rebased to current `dev`

**Pull request merged**